### DISCUSSION TITLE
How to check if querySelector contains valid CSS selector

### DISCUSSION THREAD

**Discussion by Mikilll94** [eslint/eslint#17125] [Q&A] [answered]
I would like to check if `document.querySelector` contains valid CSS selector as a first argument

```js
// This should pass
document.querySelector('#foo');

// This should fail
document.querySelector('[type="submit"');
document.querySelector('#32131');
```

Unfortunately I have not found any ESLint plugins which implement this check. Does anyone have any solution for that?

**Comment by DMartens** [comment, marked as answer]
You can use a [custom rule](https://eslint.org/docs/latest/extend/custom-rules):
Rough rule sketch to get you started:
```javascript
import { parse } from 'css-what';

export default {
	meta: {
		messages: {
			invalid: 'This CSS selector is invalid as {{ message }}',
		}
	},
	create(context) {
		return {
			// Only matches "document.querySelector(selector)"
			'CallExpression[callee.type="MemberExpression"][callee.object.name="document"][callee.property.name="querySelector"][callee.computed=false][callee.optional=false][arguments.length=1]'(call) {
				const [selector] = call.arguments;
				const isStringLiteral = selector.type === 'Literal' && typeof selector.value === 'string';
				if (!isStringLiteral) {
					return;
				}

				try {
					parse(selector);
				} catch (error) {
					context.report({
						messageId: 'invalid',
						node: selector,
						data: { message: error.message },
					})
				}
			}
		}
	}
}
```

Some notes:
- You may want to use another package for checking whether the CSS selector is invalid but you should not manually implement this because of the complexity
- This matches only `document.querySelector`, NOT `element.querySelector` or `document.querySelectorAll` (adjust the selector for this)
- The rule is propably not implemented as querySelector(All) throws an error for invalid selectors anyway

**Comment by Mikilll94** [comment]
thx @DMartens for your help :)

However I don't understand one thing. Where is the documentation for this syntax `'CallExpression[callee.type="MemberExpression"]...`? I don't see it in the link which you provided.

**Reply by DMartens** [reply]
The reference is kind of hidden in "If a key is a node type or a [selector](https://eslint.org/docs/latest/extend/selectors), ESLint calls that visitor function while going down the tree.": https://eslint.org/docs/latest/extend/selectors.
eslint uses [esquery](https://github.com/estools/esquery) for matching this selector.
This is a preference kind of thing, you can also simply use a simple selector like "CallExpression" and match manually.
You can use the [typescript-eslint playground](https://typescript-eslint.io/play/) for live filtering (you can query the AST in the top right when using "AST Viewer": ESTree)
